syntax = "proto3";

package TW.EthereumAbi.Proto;
option java_package = "wallet.core.jni.proto";

import "Common.proto";

// Generic number parameter for all other bit sizes, like UInt24, 40, 48, ... 248.
// For predefined sizes (8, 16, 32, 64, 256) use a special type like `ParamValue::number_u16`.
message NumberNParam {
    // Count of bits of the number.
    // 0 < bits <= 256, bits % 8 == 0
    uint32 bits = 1;

    // Serialized big endian.
    bytes value = 2;
}

// A fixed-size byte array.
message ByteArrayFixParam {
    // Size of the fixed array in bytes.
    uint64 size = 1;

    // Raw bytes.
    bytes value = 2;
}

// A byte array of arbitrary size.
message ArrayParam {
    repeated ParamValue values = 1;
}

// A tuple with various parameters similar to a structure.
message TupleParam {
    repeated ParamValue params = 1;
}

// An enumeration of possible parameters.
message ParamValue {
    oneof param {
        // Simple values.
        bool boolean = 1;
        uint32 number_u8 = 2;
        uint32 number_u16 = 3;
        uint32 number_u32 = 4;
        uint64 number_u64 = 5;
        bytes number_u256 = 6;
        // Gap in field numbering is intentional.
        NumberNParam number_uint = 9;
        int32 number_i8 = 10;
        int32 number_i16 = 11;
        int32 number_i32 = 12;
        int64 number_i64 = 13;
        bytes number_i256 = 14;
        // Gap in field numbering is intentional.
        NumberNParam number_int = 17;
        string string_value = 18;
        string address = 19;
        bytes byte_array = 20;
        ByteArrayFixParam byte_array_fix = 21;

        // Nested values. Gap in field numbering is intentional.
        ArrayParam array = 24;
        TupleParam tuple = 25;
    }
}

// A named parameter.
message NamedParam {
    // Name of the parameter.
    string name = 1;

    // Value of the parameter.
    ParamValue value = 2;
}

//// TWEthereumAbiDecodeContractCall

// Decode a contract call (function input) according to the given ABI json.
message ContractCallDecodingInput {
    // An encoded smart contract call with a prefixed function signature (4 bytes).
    bytes encoded = 1;

    // A smart contract ABI in JSON.
    // Each ABI function must be mapped to a short signature.
    // Expected to be a set of functions mapped to corresponding short signatures.
    // Example:
    // ```
    // {
    //     "1896f70a": {
    //         "name": "setResolver",
    //         "inputs": [...],
    //         ...
    //     },
    //     "ac9650d8": {
    //         "name": "multicall",
    //         "inputs": [...],
    //         ...
    //     }
    // }
    // ```
    string smart_contract_abi_json = 2;
}

message ContractCallDecodingOutput {
    // Human readable json format, according to the input `ContractCallDecodingInput::smart_contract_abi_json`.
    string decoded_json = 1;

    // Decoded parameters.
    repeated NamedParam params = 2;

    // error code, 0 is ok, other codes will be treated as errors
    Common.Proto.SigningError error = 3;

    // error code description
    string error_message = 4;
}

//// TWEthereumAbiDecodeParams

// Decode a function input or output data according to the given ABI json.
message ParamsDecodingInput {
    // An encoded ABI.
    bytes encoded = 1;

    // A set of ABI parameters in JSON.
    // Expected to be a JSON array at the entry level.
    // Example:
    // ```
    // [
    //     {
    //         "name": "_to',
    //         "type": "address"
    //     },
    //     {
    //         "name": "_value",
    //         "type": "uint256"
    //     }
    // ]
    // ```
    string params_abi_json = 2;
}

message ParamsDecodingOutput {
    // Decoded parameters.
    repeated NamedParam params = 1;

    // error code, 0 is ok, other codes will be treated as errors
    Common.Proto.SigningError error = 2;

    // error code description
    string error_message = 3;
}

//// TWEthereumAbiDecodeValue

// Decode an Eth ABI value.
message ValueDecodingInput {
    // An encoded value to be decoded.
    bytes encoded = 1;

    // A type of the parameter.
    string param_type = 2;
}

message ValueDecodingOutput {
    // Decoded parameters.
    ParamValue param = 1;

    // error code, 0 is ok, other codes will be treated as errors
    Common.Proto.SigningError error = 2;

    // error code description
    string error_message = 3;
}

//// TWEthereumAbiEncodeFunction

// Encode a function call to Eth ABI binary.
message FunctionEncodingInput {
    string function_name = 1;

    // Parameters to be encoded.
    repeated ParamValue params = 2;
}

message FunctionEncodingOutput {
    // The function type signature.
    // Example: "baz(int32,uint256)"
    string function_type = 1;

    // An encoded smart contract call with a prefixed function signature (4 bytes).
    bytes encoded = 2;
}
